File: Python/Json_process.py
import json
from pathlib import Path
from typing import Any


def load_json(file_path: Path) -> Any:
    with file_path.open("r", encoding="utf-8") as file:
        return json.load(file)


def save_json(file_path: Path, data: Any) -> None:
    with file_path.open("w", encoding="utf-8") as file:
        json.dump(data, file, indent=2, ensure_ascii=False)
    print(f"Saved: {file_path}")


def get_nested_value(data: Any, path: str) -> Any:
    current = data
    for key in path.split("."):
        if isinstance(current, dict) and key in current:
            current = current[key]
        else:
            raise KeyError(f"Path not found: {path}")
    return current


def main() -> None:
    file_input = input("Enter JSON file path: ").strip()
    if not file_input:
        print("File path is required.")
        return

    file_path = Path(file_input)
    if not file_path.exists():
        print("File not found.")
        return

    try:
        data = load_json(file_path)
    except json.JSONDecodeError as error:
        print(f"Invalid JSON: {error}")
        return
    except Exception as error:
        print(f"Error reading file: {error}")
        return

    print("\nPretty JSON:")
    print(json.dumps(data, indent=2, ensure_ascii=False))

    key_path = input(
        "\nEnter key path to read (example: user.name), or press Enter to skip: "
    ).strip()
    if key_path:
        try:
            value = get_nested_value(data, key_path)
            print(f'Value at "{key_path}": {value}')
        except KeyError as error:
            print(error)

    save_choice = input("\nSave pretty JSON to new file? (y/n): ").strip().lower()
    if save_choice == "y":
        output_path = file_path.with_name(f"{file_path.stem}_pretty.json")
        save_json(output_path, data)


if __name__ == "__main__":
    main()

File: Javascript/Json_process.js
const fs = require('node:fs');
const path = require('node:path');
const readline = require('node:readline/promises');
const { stdin, stdout } = require('node:process');

const loadJson = (filePath) => JSON.parse(fs.readFileSync(filePath, 'utf8'));

const saveJson = (filePath, data) => {
  fs.writeFileSync(filePath, `${JSON.stringify(data, null, 2)}\n`, 'utf8');
  console.log(`Saved: ${filePath}`);
};

const getNestedValue = (data, keyPath) => {
  let current = data;
  for (const key of keyPath.split('.')) {
    if (current && typeof current === 'object' && key in current) {
      current = current[key];
    } else {
      throw new Error(`Path not found: ${keyPath}`);
    }
  }
  return current;
};

const main = async () => {
  const rl = readline.createInterface({ input: stdin, output: stdout });

  try {
    const fileInput = (await rl.question('Enter JSON file path: ')).trim();
    if (!fileInput) {
      console.log('File path is required.');
      return;
    }

    const filePath = path.resolve(fileInput);
    if (!fs.existsSync(filePath)) {
      console.log('File not found.');
      return;
    }

    let data;
    try {
      data = loadJson(filePath);
    } catch (error) {
      console.log(`Invalid JSON: ${error.message}`);
      return;
    }

    console.log('\nPretty JSON:');
    console.log(JSON.stringify(data, null, 2));

    const keyPath = (
      await rl.question(
        '\nEnter key path to read (example: user.name), or press Enter to skip: ',
      )
    ).trim();

    if (keyPath) {
      try {
        const value = getNestedValue(data, keyPath);
        console.log(`Value at "${keyPath}":`, value);
      } catch (error) {
        console.log(error.message);
      }
    }

    const saveChoice = (
      await rl.question('\nSave pretty JSON to new file? (y/n): ')
    )
      .trim()
      .toLowerCase();

    if (saveChoice === 'y') {
      const outputPath = path.join(
        path.dirname(filePath),
        `${path.parse(filePath).name}_pretty.json`,
      );
      saveJson(outputPath, data);
    }
  } finally {
    rl.close();
  }
};

main();
